runtime/compiler/aarch64/codegen/StackCheckFailureSnippet.cpp

```diff
@@ -22,20 +22,83 @@
 
 #include "codegen/StackCheckFailureSnippet.hpp"
 
+#include "codegen/ARM64Instruction.hpp"
 #include "codegen/CodeGenerator.hpp"
+#include "codegen/GCStackAtlas.hpp"
+#include "codegen/Linkage.hpp"
+#include "codegen/Machine.hpp"
+#include "il/symbol/ResolvedMethodSymbol.hpp"
 
 uint8_t *
 TR::ARM64StackCheckFailureSnippet::emitSnippetBody()
    {
-   TR_UNIMPLEMENTED();
+   TR::ResolvedMethodSymbol *bodySymbol = cg()->comp()->getJittedMethodSymbol();
+   TR::Machine *machine = cg()->machine();
+   TR::SymbolReference *sofRef = cg()->comp()->getSymRefTab()->findOrCreateStackOverflowSymbolRef(bodySymbol);
 
-   uint8_t *buffer = cg()->getBinaryBufferCursor();
-   return buffer;
+   uint8_t *cursor = cg()->getBinaryBufferCursor();
+   uint8_t *returnLocation;
+
+   getSnippetLabel()->setCodeLocation(cursor);
+
+   // Pass cg()->getFrameSizeInBytes() to jitStackOverflow in register x9.
+   //
+   const TR::ARM64LinkageProperties &linkage = cg()->getLinkage()->getProperties();
+   uint32_t frameSize = cg()->getFrameSizeInBytes();
+
+   if (frameSize <= 0xffff)
+      {
+      // mov x9, #frameSize
+      *(int32_t *)cursor = 0xd2800009 | (frameSize << 5);
+      cursor += ARM64_INSTRUCTION_LENGTH;
+      }
+   else
+      {
+      TR_UNIMPLEMENTED();
+      }
+
+   // add J9SP, J9SP, x9
+   *(int32_t *)cursor = 0x8b090294;
+   cursor += ARM64_INSTRUCTION_LENGTH;
+
+   // str LR, [J9SP]
+   // ToDo: Skip saving/restoring LR when not required
+   *(int32_t *)cursor = 0xf900029e;
+   cursor += ARM64_INSTRUCTION_LENGTH;
+
+   // bl helper
+   *(int32_t *)cursor = cg()->encodeHelperBranchAndLink(sofRef, cursor, getNode());
+   cursor += ARM64_INSTRUCTION_LENGTH;
+   returnLocation = cursor;
+
+   // ldr LR, [J9SP]
+   *(int32_t *)cursor = 0xf940029e;
+   cursor += ARM64_INSTRUCTION_LENGTH;
+
+   // sub J9SP, J9SP, x9 ; assume that jitStackOverflow does not clobber x9
+   *(int32_t *)cursor = 0xcb090294;
+   cursor += ARM64_INSTRUCTION_LENGTH;
+
+   // b restartLabel ; assuming it is less than 28 bits away.
```

Comment:
Probably a safe bet, but rather than putting your assumption in a comment why not use an actual assert for the range of the displacement?

Reply:
Added a guard with `if (constantIsSignedImm28(distance))`.